api/src/game_state_server.py

```diff
@@ -0,0 +1,157 @@
+from dataclasses import dataclass
+
+UNIT_SIZE = 50
+
+
+@dataclass
+class Token:
+
+    id: str
+    icon_id: str
+    start_x: int
+    start_y: int
+    start_z: int
+    end_x: int
+    end_y: int
+    end_z: int
+
+    def to_dict(self):
+
+        return {
+            'id': self.id,
+            'icon_id': self.icon_id,
+            'start_x': self.start_x,
+            'start_y': self.start_y,
+            'start_z': self.start_z,
+            'end_x': self.end_x,
+            'end_y': self.end_y,
+            'end_z': self.end_z,
+        }
+
+
+class RoomData:
+
+    def __init__(self, room_id, initial_connection=None):
+
+        self.room_id = room_id
+        self.game_state = {}
+        self.id_to_positions = {}
+        self.positions_to_ids = {}
+        self.clients = set()
+        if initial_connection:
+            self.clients.add(initial_connection)
+
+
+class GameStateServer:
+
+    def __init__(self):
+
+        self._rooms = {}
+
+    def new_connection_request(self, client, room_id):
+
+        if self._rooms.get(room_id, False):
+            self._rooms[room_id].clients.add(client)
+            return self.get_state(room_id)
+        else:
+            self._rooms[room_id] = RoomData(room_id, initial_connection=client)
+            return
+
+    def connection_dropped(self, client, room_id):
+
+        if self._rooms.get(room_id, False):
+            self._rooms[room_id].clients.remove(client)
+
+    def process_updates(self, messages, room_id):
+
+        if not (self._rooms.get(room_id, False) and self._rooms[room_id].clients):
+            print('Unable to process update from new connection')
+            return
+        for message in messages:
+            action = message.get('action', None)
+            data = message.get('data', None)
+            if not action or not data:
+                print('Invalid update received')
+                return
+            token = self.convert_token(data)
+            if not token:
+                return
+            if action == 'create' or action == 'update' and \
+                    self.validate_token(token) and self.validate_position(token, room_id):
+                self.create_or_update_token(token, room_id)
+            elif action == 'delete' and data.get('id', False):
+                if self._rooms[room_id].game_state.get(data['id'], False):
+                    self.delete_token(token, room_id)
+            else:
+                print(f'Received invalid action: {action}')
+        return self.get_state(room_id)
+
+    @staticmethod
+    def convert_token(data):
+
+        try:
+            token = Token(**data)
+        except TypeError:
+            print(f'Invalid token received: {data}')
+            return None
+        return token
+
+    @staticmethod
+    def validate_token(token):
```

Comment:
Same here as below. `validate` can also mean throw an exception if it's invalid, so definitely want to be clear here